### ISSUE TITLE
Add AsyncParsableArguments to support validating async arguments

### ISSUE THREAD

**Issue by luispadron** [apple/swift-argument-parser#630]
Currently to share arguments across commands you can create a `ParsableArguments` type which contains your arguments/options, and can use `func validate() throws` to validate the arguments.

This fails for things that must be validated in an async manner. We should consider adding an async API for validating or `AsyncParsableArguments` to support this.

**Comment by CrazyFanFan** [comment]
I just encountered the same problem.

**Comment by dcantah** [comment]
Also just hit this.. thought I might've been holding it wrong. I can try and take a look

**Comment by natecook1000** [comment]
Given the headaches and challenges around supporting the sync and async versions of `run()`, I'm reticent to provide another overload here. As a workaround, you can perform any async validation at the start of your async `run()` method – if you throw a `ValidationError` from there you'll get the same output as an error thrown from the `validate()` method.

If this workaround doesn't work for you, please let me know!

**Comment by luispadron** [comment]
Yeah that works and it's what we do today but it lacks a few things:

- the argument cannot have the correct type, because the async function is not supported. E.g we need to declare the argument as String for example and then validate it converts to whatever type we wanted.
- this conversion needs to happen in all the commands which use that argument.

**Comment by ffried** [comment]
I'd like to throw another use case into the ring here:

```swift
struct PathBased {
    static func setup(withPath path: String?) async throws { /* ... */ }
}

@main
struct MainCommand: AsyncParsableCommand {
    static var configuration: CommandConfiguration {
        .init(subcommands: [
            SubCommand1.self,
        ])
    }

    @Option
    var somePath: String?

    func validate() throws {
         try await PathBased.setup(withPath: somePath) // error: 'async' call in a function that does not support concurrency 
    }
}

struct SubCommand1: AsyncParsableCommand {
   public static var configuration: CommandConfiguration {
        .init(commandName: "sub1")
    }

   public func run() async throws {
      // No way to call `PathBased.setup(withPath:)` b/c there's no way to access `MainCommand.somePath` here.
   }
}
```

It's of course possible to extract the `somePath` option into a `ParsableArguments` / `OptionGroup` sub-structure, but for one option it seems overkill.


**Update / correction:** It seems with #802 this is indeed possible. Still creates bidirectional explicit linking for something that could work out of the box.

**Comment by bhab20476-prog** [comment]
هاي

في الثلاثاء، ٣٠ أيلول ٢٠٢٥ ١٠:٤٠ ص Florian ***@***.***> كتب:
